### ISSUE TITLE
uprobe DWARF location attachment does not work with uprobe_multi

### ISSUE THREAD

**Issue by viktormalik** [bpftrace/bpftrace#3173]
#3095 introduced parsing of DWARF for a binary and attaching uprobes after the function prologue. This is currently not used in combination with uprobe_multi, compare:
```
# bpftrace  -e 'uprobe:tests/testprogs/uprobe_loop:uprobeFunction1 {  printf("%s\n", ustack); exit(); }' -c tests/testprogs/uprobe_loop
Attaching 1 probe...

        uprobeFunction1+13
        spin+37
        main+25
        __libc_start_call_main+122
        __libc_start_main_alias_1+139
        _start+37

# bpftrace  -e 'uprobe:tests/testprogs/uprobe_loop:uprobeFunction* {  printf("%s\n", ustack); exit(); }' -c tests/testprogs/uprobe_loop
Attaching 1 probe...

        uprobeFunction1+0
        main+25
        __libc_start_call_main+122
        __libc_start_main_alias_1+139
        _start+37
```
In the second stack, uprobe is not attached after the prologue.

Since uprobe_multi link actually accepts offsets rather that function names, it should be possible to use these two mechanisms together.

cc @ttreyer

**Comment by ttreyer** [comment]
> Since uprobe_multi link actually accepts offsets rather that function names, it should be possible to use these two mechanisms together.

Just to understand clearly, I could pass something like "uprobeFunction1+1234" inside the `probe.funcs` vector and the Kernel will set the probe at the specified offset?

**Comment by ajor** [comment]
It's bpftrace that does the symbol resolution already in [resolve_offset_uprobe_multi](https://github.com/bpftrace/bpftrace/blob/d592be8fe65d7b49aa183eb1a05625df1743ed0c/src/attached_probe.cpp#L1030-L1034).

Then we just pass the raw offsets to the kernel here: https://github.com/bpftrace/bpftrace/blob/d592be8fe65d7b49aa183eb1a05625df1743ed0c/src/attached_probe.cpp#L1102

**Comment by viktormalik** [comment]
> It's bpftrace that does the symbol resolution already in [resolve_offset_uprobe_multi](https://github.com/bpftrace/bpftrace/blob/d592be8fe65d7b49aa183eb1a05625df1743ed0c/src/attached_probe.cpp#L1030-L1034).
> 
> Then we just pass the raw offsets to the kernel here:
> 
> https://github.com/bpftrace/bpftrace/blob/d592be8fe65d7b49aa183eb1a05625df1743ed0c/src/attached_probe.cpp#L1102

Exactly. So, we could support storing `function+offset` in `probe.funcs` and then just add `offset` to the raw ELF offset passed to `opts.uprobe_multi.offsets`.

**Comment by ttreyer** [comment]
Is there something I need to enable in the Kernel?
I'm running `Linux 6.8.9-300.fc40.aarch64`, but `bpftrace --info` shows:
```
Probe types
  kprobe: yes
  tracepoint: yes
  perf_event: yes
  kfunc: yes
  kprobe_multi: no
  uprobe_multi: no
  raw_tp_special: yes
  iter: yes
```

**Comment by ajor** [comment]
I don't think there are any kernel settings to configure. Have you got a recent version of libbpf? Looks like it'll need 1.3 or newer: https://github.com/libbpf/libbpf/releases

**Comment by tyroguru** [comment]
I've just been doing some work with libbpf built from very recent source (HEAD == bc24cd1) and my dev system now has uprobe_multi enabled as a result. I'm trying to use the new inlined function feature and I'm specifying the member function with the `:cpp` specifier. The test case is trivial:

```
struct Foo {
...
  inline void doit(std::string& str) {
    /* stuff */
  };
};
```

The probe specification is `uprobe:/path/to/bin:cpp:"Foo::doit"`. The expansion type is set to `ExpansionType::MULTI` in `AttachPointParser::uprobe_parser()` which means we always use the symbol table to lookup symbols. This effectively renders the inline function feature unusable as a user would need to the mangled function name for the inlined symbol (although I haven't tested that) and that's not easy to obtain.

**Comment by SPYFF** [comment]
Be aware this is something might been already fixed in kernel side at least for x86: https://lore.kernel.org/all/172285896642.2215.15269623510741010803.tip-bot2@tip-bot2/

Also it might worth the take a look if #3095 works well with the fixed kernel as well. Probably yes.

**Comment by viktormalik** [comment]
> Be aware this is something might been already fixed in kernel side at least for x86: https://lore.kernel.org/all/20240729175223.23914-1-andrii@kernel.org/
> 
> Also it might worth the take a look if #3095 works well with the fixed kernel as well. Probably yes.

Thanks for quoting the above fix. It seems to be specific to perf, though, I'm not sure if we'll be able to use it.

**Comment by ajor** [comment]
> Be aware this is something might been already fixed in kernel side at least for x86: https://lore.kernel.org/all/20240729175223.23914-1-andrii@kernel.org/
> 
> Also it might worth the take a look if #3095 works well with the fixed kernel as well. Probably yes.

Andrii pointed this out to me and it is applicable to bpftrace, but it'll only be in the absolutely latest kernels and will take a while to roll out everywhere. Using LLDB to find uprobe attachment points means we'll produce correct stack traces on all kernels, new or old.

The disadvantage is that the instructions we're attaching on most likely aren't going to be emulated in the kernel, unlike the `push %rbp` which is the first instruction of most functions. This means our uprobes by default will incur some extra overhead, but I think it's worth it so that we always produce accurate data for stack traces.

It's possible to override LLDB's prologue skipping by specifying an offset of 0: `uprobe:exe:my_func+0`. If you know you're on a newer kernel you can do this to take advantage of the accurate stack traces and improved performance. It'd be good if we could auto-detect this kernel functionality and disable our use of LLDB automatically.

**Comment by ttreyer** [comment]
> It's possible to override LLDB's prologue skipping by specifying an offset of 0: uprobe:exe:my_func+0. If you know you're on a newer kernel you can do this to take advantage of the accurate stack traces and improved performance. It'd be good if we could auto-detect this kernel functionality and disable our use of LLDB automatically.

Unfortunately, this won't work. We don't distinguish between `my_func+0` and `my_func`, so in both case we'll call to LLDB in order to skip the prologue.

**Comment by ajor** [comment]
> Unfortunately, this won't work. We don't distinguish between `my_func+0` and `my_func`, so in both case we'll call to LLDB in order to skip the prologue.

I've opened an issue so we don't forget this: #3469. Don't think it's a major problem to be fixed immediately though.
